cli/flox-rust-sdk/src/providers/services.rs

```diff
@@ -119,8 +186,168 @@ pub fn maybe_make_service_config_file(
+    status: String,
+    system_time: String,
+    age: u64,
+    is_ready: String,
+    restarts: u64,
+    exit_code: i32,
+    pid: u64,
+    #[serde(rename = "IsRunning")]
+    is_running: bool,
+}
+
+#[derive(Deserialize, Debug, Clone, PartialEq)]
+pub struct ProcessStates(Vec<ProcessState>);
+
+impl ProcessStates {
+    pub fn read(socket: impl AsRef<Path>) -> Result<ProcessStates, ServiceError> {
+        let mut cmd = base_process_compose_command(socket.as_ref());
+        let output = cmd
+            .arg("list")
+            .args(["--output", "json"])
+            .output()
+            .map_err(ServiceError::ProcessComposeCmd)?;
+        if !output.status.success() {
+            return Err(ServiceError::from_output(String::from_utf8_lossy(
+                &output.stderr,
+            )));
+        }
+        let mut processes: ProcessStates =
+            serde_json::from_slice(&output.stdout).map_err(ServiceError::ParseOutput)?;
+        processes
+            .0
+            .retain(|state| state.name != PROCESS_NEVER_EXIT_NAME);
+
+        Ok(processes)
+    }
+
+    pub fn get_running_names(&self) -> Vec<String> {
+        self.0
+            .iter()
+            .filter(|state| state.is_running)
+            .map(|state| state.name.clone())
+            .collect()
+    }
+}
+
+/// Constructs a base `process-compose process` command to which additional
+/// arguments can be appended.
+fn base_process_compose_command(socket: impl AsRef<Path>) -> Command {
+    let path = Path::new(&*PROCESS_COMPOSE_BIN);
+    let mut cmd = Command::new(path);
+    cmd.env("PATH", path)
+        .env("NO_COLOR", "1") // apparently it doesn't do this automatically even though it's not connected to a tty...
+        .arg("--unix-socket")
+        .arg(socket.as_ref().to_string_lossy().as_ref())
+        .arg("process");
+
+    cmd
+}
+
+/// Stop service(s).
+pub fn stop_services(
+    socket: impl AsRef<Path>,
+    names: &[impl AsRef<str>],
+) -> Result<(), ServiceError> {
+    let names = names.iter().map(|name| name.as_ref()).collect::<Vec<_>>();
+    tracing::debug!(names = names.join(","), "stopping services");
+
+    let mut cmd = base_process_compose_command(socket);
+    let output = cmd
+        .arg("stop")
+        .args(names)
+        .stderr(Stdio::piped())
+        .stdout(Stdio::piped())
+        .output()
+        .map_err(ServiceError::ProcessComposeCmd)?;
+
+    if output.status.success() {
+        tracing::debug!("services stopped");
+        Ok(())
+    } else {
+        tracing::debug!("stopping services failed");
+        let stderr = String::from_utf8_lossy(&output.stderr);
+        Err(ServiceError::from_output(stderr))
+    }
+}
+
+/// Error message extracted from process-compose logs
+#[derive(Debug, Clone)]
+pub struct ProcessComposeLogContents {
+    pub err_msg: String,
+    pub cause_msg: String,
+}
+
+/// The types of errors that are logged by process-compose
+#[derive(Debug, Clone, PartialEq, thiserror::Error)]
```

Comment:
nonblocking: it's confusing for me to have `LoggedError`, `ProcessComposeLogContents`, and `impl ServiceError` without any comments explaining the difference

Reply:
Added some docs.